### PR TITLE
[Merged by Bors] - feat: batch ballots

### PR DESCRIPTION
**Pull request by acud**
## Motivation

Batches ballots to be written together into the db, reducing the amount of time we spend writing ballots to the db.

## Description

Adds a component to `mesh` that orchestrates the write coalescing of ballots.

The main problem encountered here is what happens when a node may have multiple ballots received and processed at the same time. Those might be a duplicate of the same ballot received in the same time (theoretical, not sure if it can really happen) and malfeasant nodes (different ballots at the same time).

This can result in multiple ballots put into the same batch, circumventing the malfeasance proof generation step in storing the batch. 

I've overcome this by keeping tabs over the identities of smeshers that are included in a batch to flush, and allowing only one ballot per identity per batch. In this case, the caller gets a retry function that needs to be called (and returns once the batch is flushed), allowing the caller to try to write that same ballot again.

Another side-effect here is that the lock over the write-coalescer must be held during the entire time while a batch is being written. I've thought to go in the approach of a CoW here but in this case it meant that I'd have to keep tabs on the identities both in the batch being written, _and_ the next batch which will be written, and I figured this is just a little too much overhead and too much complexity for a first iteration.

## Test Plan

- [x] check performance gain on prod

## TODO

- [x] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] ~~Update documentation as needed~~
- [x] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: metrics/common.go [hunk 1]
```diff
@@ -76,9 +83,15 @@ func ReportMessageLatency(protocol, msgType string, latency time.Duration) {
...
+// NewSimpleCounter creates a simple prometheus counter.
+func NewSimpleCounter(subsystem, name, help string) prometheus.Counter {
+	return promauto.NewCounter(NewCounterOpts(subsystem, name, help))
+}
+
 func NewCounterOpts(ns, name, help string) prometheus.CounterOpts {
 	return prometheus.CounterOpts{
-		Namespace: ns,
+		Namespace: Namespace,
+		Subsystem: ns,
```
Comment (poszu) [inline metrics/common.go]: Good catch :) How about naming the variable for the subsystem accordingly instead of `ns`?

File: mesh/mesh_test.go [hunk 2]
```diff
@@ -66,6 +66,9 @@ func createTestMesh(t *testing.T) *testMesh {
 	exec := NewExecutor(db, atxsdata, tm.mockVM, tm.mockState, lg)
 	msh, err := NewMesh(db, atxsdata, tm.mockClock, tm.mockTortoise, exec, tm.mockState, lg)
 	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+	go msh.Start(ctx)
```
Comment (poszu) [inline mesh/mesh_test.go]: I think the test should wait for `msh` to stop to avoid bugs like logging after the test has already finished.

File: mesh/mesh.go [hunk 3]
```diff
@@ -551,61 +560,48 @@ func (msh *Mesh) AddBallot(
...
+	msh.addBallotMu.Unlock()
+	start := time.Now()
+	if retry != nil {
+		// this means that we got a ballot from the same smesher in this layer.
+		// while this doesn't directly implicate a node as being malfeasant, it does
+		// force us to retry this operation once the current batch is written to the db
+		retry()
+		metrics.BallotWaitRetry.Observe(time.Since(start).Seconds())
+		goto START
+	}
```
Comment (poszu) [inline mesh/mesh.go]: The `goto` could be easily avoided by using a loop:
```suggestion
	var (
		initiallyMalfeasant bool
		start               time.Time
		waitC               <-chan struct{}
		errFunc             func() error
	)
	for {
		initiallyMalfeasant = false
		msh.addBallotMu.Lock()
		malicious := msh.atxsdata.IsMalicious(ballot.SmesherID)
		if malicious {
			ballot.SetMalicious()
			initiallyMalfeasant = true
		}

		// Store should only allow one ballot per node ID per batch.
		var retry func()
		waitC, errFunc, retry = msh.ballotWriter.Store(ballot)
		msh.addBallotMu.Unlock()
		start = time.Now()
		if retry == nil {
			break
		}
		// this means that we got a ballot from the same smesher in this layer.
		// while this doesn't directly implicate a node as being malfeasant, it does
		// force us to retry this operation once the current batch is written to the db
		retry()
		metrics.BallotWaitRetry.Observe(time.Since(start).Seconds())
	}

```

File: mesh/mesh.go [hunk 4]
```diff
@@ -551,61 +560,48 @@ func (msh *Mesh) AddBallot(
...
+	msh.addBallotMu.Lock()
 	malicious := msh.atxsdata.IsMalicious(ballot.SmesherID)
 	if malicious {
 		ballot.SetMalicious()
+		initiallyMalfeasant = true
+	}
+
+	// Store should only allow one ballot per node ID per batch.
+	waitC, errFunc, retry := msh.ballotWriter.Store(ballot)
+	msh.addBallotMu.Unlock()
```
Comment (poszu) [inline mesh/mesh.go]: Does `msh.atxsdata.IsMalicious()` must be in the locked section? How about:
```suggestion
	malicious := msh.atxsdata.IsMalicious(ballot.SmesherID)
	if malicious {
		ballot.SetMalicious()
		initiallyMalfeasant = true
	}

	// Store should only allow one ballot per node ID per batch.
	msh.addBallotMu.Lock()
	waitC, errFunc, retry := msh.ballotWriter.Store(ballot)
	msh.addBallotMu.Unlock()
```

File: mesh/ballotwriter/ballotwriter.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+			return
+		case <-w.timer.C:
+			// we hang on to this lock for the entire duration of this select case branch.
+			// this is in order to simplify the logic when handling multiple writes from the
+			// same smesher ID. Implementing copy-on-write here comes at a big taxation because
+			// one needs to both keep track of the identities in the batch being currently
+			// written, as well as the one which would be queued up for the next write.
+			w.atxMu.Lock()
+			startTime := time.Now()
+			w.running = false
```
Comment (poszu) [inline mesh/ballotwriter/ballotwriter.go]: The `w.running` seems to be unused. It's only set to false here and its never read.

File: mesh/ballotwriter/ballotwriter.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+			}
+			w.rmNodeSlice = w.rmNodeSlice[:0]
+
+			for k := range batch {
+				w.rmSlice = append(w.rmSlice, k)
+			}
+			for _, k := range w.rmSlice {
+				delete(batch, k)
+			}
+			w.rmSlice = w.rmSlice[:0]
```
Comment (poszu) [inline mesh/ballotwriter/ballotwriter.go]: If I understand correctly, you want to clear `w.ballotBatchIdentities` and `w.ballotBatch`. How about:
```suggestion
			clear(w.ballotBatchIdentities)
			clear(w.ballotBatch)
```

and remove the `rmSlice` and `rmNodeSlice` fields from the `BallotWriter` struct

File: mesh/ballotwriter/ballotwriter.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+			CleanupTime.Add(float64(time.Since(cleanupStart)))
+			close(res.doneC)
+		}
+	}
+}
+
+// Store a ballot. Returns a channel to wait for the result, a function that returns
+// the error value and a wait function to try again. The retry logic can happen when
+// a node has multiple ballots on the same layer (malfeasant node or duplicate ballot received).
+func (w *BallotWriter) Store(b *types.Ballot) (<-chan struct{}, func() error, func()) {
```
Comment (poszu) [inline mesh/ballotwriter/ballotwriter.go]: I think it would be much easier if this method were blocking. It can wait for the batch to be submitted internally, wdyt?
```suggestion
func (w *BallotWriter) Store(ctx context.Context, b *types.Ballot) (error) {
```

**Review by poszu — CHANGES_REQUESTED** [review]

File: mesh/ballotwriter/ballotwriter.go (hunk 6 above)
Comment (acud) [reply mesh/ballotwriter/ballotwriter.go]: nice. I actually didn't know this builtin

File: mesh/mesh.go (hunk 4 above)
Comment (acud) [reply mesh/mesh.go]: The motivation behind putting it inside the lock was the following:
- on the writer side (here) we want to check whether the identity is malicious and set the ballot as malicious before adding it, under lock, because:
- on the consumer side (in the forever loop `Start`) we hold the lock during the whole execution time of writing the batch, and during that time, an identity might turn into a malicious one


This is also why I eventually went away from doing a CoW for the new batch (allowing to queue writes while another batch is already being written) among other reasons (also because of the complexity of keeping tabs over the single-ballot-per-identity-per-batch part).
So TL;DR this is done this way because of at least a hypothetical data race I was trying to avoid. That being said, an identity can _still_ turn into a malicious one for other reasons during that time due to other protocols.
I'm not sure if this makes too much sense though, let me know your thoughts on this. Happy to make the proposed change.

Comment (acud) [reply mesh/mesh.go]: Ok actually given this some thought and the mutex here might not be at all needed because of the locking strategy inside the ballot writer. Ideally, the action of writing the ballot and setting the identity as malicious would be atomic but I'm not sure that's achievable.

File: mesh/ballotwriter/ballotwriter.go (hunk 7 above)
Comment (acud) [reply mesh/ballotwriter/ballotwriter.go]: I have a slight aversion to downstream components having their own retry logic. I generally try to avoid this by policy and allow the calling context to do retries (it also usually just makes reading the code and keeping a mental map of it more difficult). 

It will, however, simplify the retry logic from the calling context; agreed. I am generally happy with it as-is, with the only caveat that I now realize there's a missing select case on the for loop that checks for an expired context before calling `Store` (this, however, is likely non-fatal because it will just queue the value for storing, later on exiting on a context cancelled, unless there's a retry in progress).

File: mesh/mesh.go (hunk 4 above)
Comment (acud) [reply mesh/mesh.go]: I ended up removing the mutex

File: mesh/ballotwriter/ballotwriter.go (hunk 7 above)
Comment (poszu) [reply mesh/ballotwriter/ballotwriter.go]: Making this method blocking doesn't mean that the caller cannot retry. It can decide whether to retry or not based on the returned error. But, why would the caller want to retry? Which failures are retriable?

Comment (poszu) [reply mesh/ballotwriter/ballotwriter.go]: I looked at the code again to understand when the `retry` is used. It's used to avoid an ID putting more than 1 ballot into a batch to "avoid problems with malfeasance detection" (double ballot publish). I think this is not needed. If an ID puts 2 ballots for the same layer in a batch, then the existing malfeasance detection logic will detect it. Precisely, the following logic will catch it:

```go
db.WithTx(context.Background(), func(tx sql.Transaction) error {
	require.NoError(t, ballots.Add(tx, ballot))
	// check if ballot is found
	existing, err := ballots.LayerBallotByNodeID(tx, ballot.Layer, ballot.SmesherID)
	// err == nil
	// existing == ballot
}
```

The following test passes:
```go
func TestMalfeasanceDetectionInBatch(t *testing.T) {
	db := statesql.InMemoryTest(t)
	ballot := types.NewExistingBallot(
		types.BallotID{1, 2, 3},
		types.EmptyEdSignature,
		types.RandomNodeID(),
		77,
	)
	db.WithTx(context.Background(), func(tx sql.Transaction) error {
		require.NoError(t, ballots.Add(tx, &ballot))

		// check if ballot is found
		existing, err := ballots.LayerBallotByNodeID(tx, ballot.Layer, ballot.SmesherID)
		require.NoError(t, err)
		require.Equal(t, ballot, *existing)
		return nil
	})
}
```

Therefore I think the `BallotWriter::Store()` can safely put the ballot into the batch without checking the "dirty" set `w.ballotBatchIdentities`. I suggest to change this method to:

```go
func (w *BallotWriter) Store(ctx context.Context, b *types.Ballot) error {
	w.atxMu.Lock()
	br := w.ballotBatchResult
	w.ballotBatch[b.ID()] = b
	w.atxMu.Unlock()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-br.doneC:
		return br.err
	}
}
```

Comment (acud) [reply mesh/ballotwriter/ballotwriter.go]: Yeah so the idea was that one smesher ID cannot write more than one ballot per batch because I didn't want to add the complexity of checking whether a ballot exists for the smesher in the batch, and if yes, check the ID of the ballot, compare against the other, and generate a malfeasance proof as a result (optionally). I felt this is such a rare case it does not worth the effort of having this amount of extra logic - better just check if an entry exists for the identity in the batch and basically tell the caller to back-off on the request instead, trying again later.
To me there are possibly two different scenarios:
- the same ballot arriving twice by gossip or sync from different peers (I'm not sure if this is possible but I wanted to guard against it anyway) at the same time. this is non-fatal and should be idempotent anyway
- two different ballots from the same peer come at the same time (creating a malfeasance proof)
  - these two end up on the same batch
  - if we don't check the identity limitation (one-per-identity-per-batch), the main question is - if there's two ballots and I write them within the domain of the transaction, when the transaction does a query after "inserting"(but not committing) one of them, will it see that same uncommitted ballot when querying the table, if the transaction is not yet committed?

Comment (poszu) [reply mesh/ballotwriter/ballotwriter.go]: > the main question is - if there's two ballots and I write them within the domain of the transaction, when the transaction does a query after "inserting"(but not committing) one of them, will it see that same uncommitted ballot when querying the table, if the transaction is not yet committed?

My claim is that it will, which the test above shows. The additional check and back off mechanism is not needed. You could modify the retry test to verify this.

Comment (acud) [reply mesh/ballotwriter/ballotwriter.go]: Ok, point made about not needing to keep track over identities, however, passing the context that you want to pass is incorrect, simply because the forever loop of the write-coalescer has a different context (which is cancelled on shutdown).

The problem here is that if I pass the context which is propagated from the request, we might cancel the context but still end up with the specific ballot being written to the database which I find to be counter-intuitive and probably also leading to some inconsistencies. Would still vote to keep this without passing the context to the `Store` method.

File: mesh/mesh.go [hunk 8]
```diff
@@ -551,61 +559,32 @@ func (msh *Mesh) AddBallot(
...
+	var (
+		initiallyMalfeasant bool
+		start               time.Time
+	)
+	initiallyMalfeasant = false
 	malicious := msh.atxsdata.IsMalicious(ballot.SmesherID)
 	if malicious {
 		ballot.SetMalicious()
+		initiallyMalfeasant = true
 	}
```
Comment (poszu) [inline mesh/mesh.go]: NIT: `initiallyMalfeasant` isn't needed anymore:
```suggestion
	malicious := msh.atxsdata.IsMalicious(ballot.SmesherID)
	if malicious {
		ballot.SetMalicious()
	}

	start := time.Now()
	// use `malicious` in place of `initiallyMalfeasant`
```

**Review by poszu — APPROVED** [review]

File: mesh/mesh.go (hunk 8 above)
Comment (acud) [reply mesh/mesh.go]: applied via extra commit. thanks for the thorough review @poszu

**Comment by acud** [comment]
bors merge

**Comment by acud** [comment]
bors merge

**Comment by acud** [comment]
bors merge

File: mesh/ballotwriter/ballotwriter.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+)
+
+var writerDelay = 100 * time.Millisecond
+
+type BallotWriter struct {
+	db     db
+	logger *zap.Logger
+
+	atxMu sync.Mutex
```
Comment (fasmat) [inline mesh/ballotwriter/ballotwriter.go:26]: `atxMu` doesn't quite fit as name. Maybe `batchMtx`?

File: mesh/ballotwriter/ballotwriter.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+type batchResult struct {
+	doneC chan struct{}
+	err   error
+}
+
+type db interface {
+	WithTx(context.Context, func(sql.Transaction) error) error
+	sql.Executor
+}
```
Comment (fasmat) [inline mesh/ballotwriter/ballotwriter.go:170]: I think this could be replaced with the `sql.Database` interface to avoid having a separate definition for interface used.

File: mesh/ballotwriter/ballotwriter_test.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/spacemeshos/go-spacemesh/sql/ballots"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+	"github.com/spacemeshos/go-spacemesh/sql/statesql"
+	"github.com/spacemeshos/go-spacemesh/sql/statesql/migrations"
+)
+
+func init() {
+	// otherwise we get a divide by zero error
+	types.SetLayersPerEpoch(1)
+}
```
Comment (fasmat) [inline mesh/ballotwriter/ballotwriter_test.go:29]: It would be better to put that into `TestMain` otherwise this might pollute the global state of the application.

File: mesh/mesh_test.go [hunk 12]
```diff
@@ -66,6 +66,14 @@ func createTestMesh(t *testing.T) *testMesh {
...
 	msh, err := NewMesh(db, atxsdata, tm.mockClock, tm.mockTortoise, exec, tm.mockState, lg)
 	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	done := make(chan struct{})
+	go func() {
+		msh.Start(ctx)
+		close(done)
+	}()
+	t.Cleanup(func() { <-done })
+	t.Cleanup(cancel)
```
Comment (fasmat) [inline mesh/mesh_test.go:76]:
```suggestion
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	var eg errgroup.Group
	eg.Go(func() error {
		mesh.Start(ctx)
		return nil
	})
	t.Cleanup(func() { assert.NoError(t, eg.Wait()) } )
```

**Review by fasmat — COMMENTED** [review]
I know I'm late for review but I added a few smaller suggestions.

File: mesh/mesh_test.go (hunk 12 above)
Comment (acud) [reply mesh/mesh_test.go:76]: the error group will never return an error so why is the assert needed?

Comment (fasmat) [reply mesh/mesh_test.go:76]: Yeah, it's not. Just a force of habit from other tests 🙂
